### ISSUE TITLE
Negative Efficiency SCF Gradient Threading

### ISSUE THREAD

**Issue by JonathonMisiewicz** [psi4/psi4#2559]
Carried forward from [forum request](http://forum.psicode.org/t/calculations-with-custom-basis-sets-run-slower-with-multiple-threads/2461).

According to the reporter, upon using a custom gbs file (provided in the issue), increasing the number of threads causes an SCF gradient computation to take _longer_ on a system as simple as water.

I haven't reproduced this yet, but this is serious if true.

**Comment by susilehtola** [comment]
Looking at the output in the forum, a big difference is
```
Libint2ERI::Libint2ERI              :   1268.900u      0.767s    763.645w      7 calls
```
one one thread vs
```
Libint2ERI::Libint2ERI              :   5577.683u     13.983s   3359.746w     42 calls
```
on 8 threads.

**Comment by zachglick** [comment]
I agree @susilehtola , the problematic code is [here](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/scfgrad/jk_grad.cc#L308-L310) and [here](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/scfgrad/jk_grad.cc#L883-L888)

What's happening is this: the JKGrad code creates a `TwoBodyAOInt` object per thread, in serial. The extra overhead in creating 8 vs 1 of these objects far outweighs any parallelism gained in the use of additional threads.

I suspect this performance issue was introduced by the libint2 ERI overhaul, which happened in psi4 v1.4. If I recall correctly, as part of the overhaul, the `TwoBodyAOInt` constructor was changed to automatically perform Schwarz/CSAM screening and store a list of significant shell pairs. Before the overhaul, such a list of significant shell pairs was determined outside of the `TwoBodyAOInt` object, using an `ERISieve` object. As a consequence of this overhaul, the `TwoBodyAOInt` constructor is now very expensive (because it computes all `(mn|mn)` shell quartets to do the screening), and `TwoBodyAOInt` objects should be initialized sparingly.

When many identical `TwoBodyAOInt` objects are needed, we should instead construct a single object and clone the rest, as is done [here](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/dlpno/mp2.cc#L635-L638) for example.

 Hopefully @andysim can confirm this.

**Comment by JonathonMisiewicz** [comment]
Thanks for the investigation.

[Conventional integrals for correlated code](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/deriv.cc#L548-L550) would have the same problem. As does the correlated [DF](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/mintshelper.cc#L2169-L2172) [code](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/mintshelper.cc#L2279-L2281).

**Comment by susilehtola** [comment]
The initialization of the Schwartz/CSAM screening can also be done in parallel, since it's an embarrassingly parallel `O(Nshells^2)` loop.

**Comment by zachglick** [comment]
Yeah, we need to add threading to [this](https://github.com/psi4/psi4/blob/c18efeda5af93d965314366774e401e7a0c76453/psi4/src/psi4/libmints/twobody.cc#L227) function

**Comment by JonathonMisiewicz** [comment]
Please move all further discussion of adding threading of the screening to above issue, and keep this issue about the extra work done in the threaded case.

**Comment by JonathonMisiewicz** [comment]
Fixed by #2581.
